client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -150,17 +179,19 @@ private final class PoolManager[F[_], A <: Connection[F]](
 
             case None if waitQueue.isEmpty =>
               logger.debug(s"Returning idle connection to pool: $stats")
-              idleQueue.enqueue(connection)
+              val q = idleQueue.getOrElse(key, mutable.Queue.empty[A])
+              q.enqueue(connection)
+              idleQueue.update(key, q)
 
             // returned connection didn't match any pending request: kill it and start a new one for a queued request
             case None =>
               connection.shutdown()
-              allocated -= 1
-              val Waiting(key, callback) = waitQueue.dequeue()
-              createConnection(key, callback)
+              decrConnection(key)
+              val Waiting(k, callback) = waitQueue.dequeue()
+              createConnection(k, callback)
```

Comment:
This could end up breaking the `maxConnectionsPerRequestKey` invariant since a borrow request could have ended up in `waitQueue` due to exceeding the per-key connection limit, not the `maxTotal` limit.

Reply:
We check again in createConnection for that, right?

Reply:
Yes, you're right: we won't break the invariant. If the head could be blocked on its `maxConnectionsPerRequestKey` so it will be put back at the rear of the queue while a different key may be able to open a connection.

Reply:
I just double checked the code of `createConnection` and I believe my original comment stands: it will double checks the invariant but as its written its not intended to handle that situation: it just screams very loudly about it and closes the pending dispatch with an error.

Reply:
So are you saying the invariant is broken? I still don't see how :(

Reply:
Lets say `maxConnectionsPerRequestKey` always returns 2.

1. We make one request to service A, which does not complete yet. 
1. We then make three requests to service B, two make new connections and one gets queued in `waitQueue`. The active requests do not complete yet.
1. The request to service A finishes, and finds its way to this branch, and ends up breaking the invariant since it attempts to make a new connection to service B when there are already two outstanding connections.

I believe this problem also affects the non-recycleable branch of the `release` method.

Reply:
> The request to service A finishes, and finds its way to this branch, and ends up breaking the invariant since it attempts to make a third connection to service B when there are already two outstanding connections.

But it will fail, right? CreateConnection will successfully check the condition and complete with failure.

Reply:
Yes, it will fail, but that isn't desired behavior: we should leave the third request for B in the queue until one of the first two request to B finishes. The failure pathway of `createConnection` is best described as a verbose assert: its not intended to be exercised, only to let everyone know that things aren't working as we thought. Thats why the error message is so agressive:
```scala
s"Invariant broken in ${this.getClass.getSimpleName}! Tried to create more connections than allowed: $stats"
```

Reply:
Ok, I thought you were saying a new connection will be made breaking the max connections invariant.

Reply:
I removed that entire case, it wouldn't be reached in any condition.

Reply:
I think that specific problem is still there and we have another interesting problem: it's possible for other keys to be holding idle connections summing to `maxTotal` and now we won't close one of those for a different key. That used to be handled on what was line 102.

Reply:
Which idleQueue do we dequeue from? There is no total ordering anymore. We could randomly select a queue, would that be good enough?

Reply:
Good question. I see a few options.
1. Do what you just suggested and just close a random one
1. Keep all idle connections in the same queue
1. Use a more complex data structures like LinkedHashMap

Hard to say which is ideal, and it likely depends on the circumstances. I'd probably start with 2 myself.

Reply:
LinkedHashMap would not be the same semantically, since it will order by insertion of keys (hostNames) and not the actual connections. 

I was playing with the idea of a Map pointing into a single queue but I thought that would be overkill and "choose random, dequeue" would be a better solution. I thing random dequeue is fine, what do you think?

Reply:
Sorry, `LinkedHashMap` was a knee jerk reaction to assuming that java didn't have a `LinkedHashSet`: basically yet-another-collection to store the order of idleness. I think closing a random idle connection is probably fine until proven otherwise and simpler to implement although tougher to test.

Reply:
I believe this is still broken: we need to skim the waitQueue for a waiting request that is not waiting on `maxConnectionsPerRequestKey`.

Reply:
I think @bryce-anderson is correct.  I think we want to do something like:

```scala
waitQueue.dequeueFirst { 
  waiter => allocated.getOrElse(waiter.key) < maxConnectionsPerKey(waiter.key)
} match {
  case Some(Waiter(key, cb)) =>
    // This is the first waiter not blocked on the request key limit.
    // Close the undesired connection and wait for another.
    connection.shutdown()
    decrConnection(key)
    createConnection(k, callback)
  case None =>
     // We're blocked not because of too many connections, but
     // because of too many connections per key.  We might be
     // able to reuse this request.
     idleQueues.add(connection.key, connection)
}
```